### DISCUSSION TITLE
Assigning solver options using Python.

### DISCUSSION THREAD

**Discussion by hbshrestha** [ERGO-Code/HiGHS#2051] [Announcements]
Hi. I am a complete beginner to using the HIGHS solver.
I am using it to solve a mps file using Python.

I have a very simple script, inspired from your [documentation](https://ergo-code.github.io/HiGHS/dev/interfaces/python/example-py/).

```
import highspy
import time

h = highspy.Highs()
#read model
filename = 'abc.mps'
h.readModel(filename)

#run the model
h.run()
print('Model ', filename, ' has status ', h.getModelStatus())

solution = h.getSolution()
basis = h.getBasis()
info = h.getInfo()
model_status = h.getModelStatus()
print('Model status = ', h.modelStatusToString(model_status))
print()
print('Optimal objective = ', info.objective_function_value)
print('Iteration count = ', info.simplex_iteration_count)
print('Primal solution status = ', h.solutionStatusToString(info.primal_solution_status))
print('Dual solution status = ', h.solutionStatusToString(info.dual_solution_status))
print('Basis validity = ', h.basisValidityToString(info.basis_validity))

#write solution
h.writeSolution("abc_solution.sol",1)
```

I have a few questions in this regard:
1. I read [here](https://ergo-code.github.io/HiGHS/dev/solvers/) that it is possible to use the interior point method for solving the problem. Is the default method simplex? What is the syntax for it? I tried `h.run(solver = "ipm")`, but I get TypeError with it.

2. How can I set the `parallel` [parameter](https://ergo-code.github.io/HiGHS/dev/options/definitions/#parallel) in the above Python script? For ipm, is it off by default?

3. How can I set `run_crossover` [parameter](https://ergo-code.github.io/HiGHS/dev/options/definitions/#run_crossover) to off in the above Python script? I see the crossover is set to on as default.

4. How can I set `infinite_cost` [parameter](https://ergo-code.github.io/HiGHS/dev/options/definitions/#infinite_cost) and `infinite_bound` [parameter](https://ergo-code.github.io/HiGHS/dev/options/definitions/#infinite_bound) using Python?

**Comment by jajhall** [comment]
> I have a few questions in this regard:
> 
>     1. I read [here](https://ergo-code.github.io/HiGHS/dev/solvers/) that it is possible to use the interior point method for solving the problem. Is the default method simplex? What is the syntax for it? I tried `h.run(solver = "ipm")`, but I get TypeError with it.

You can't set options this way, although there is some argument for `highspy`'s `run` to take the same arguments as the command line HiGHS executable.

You need to make an explicit call to `setOptionValue("solver", "ipm")`, as illustrated in

https://github.com/ERGO-Code/HiGHS/blob/4a7f24ac63b3832e30694969e0a69a4b3c1f1674/examples/call_highs_from_python_highspy.py#L29


>     2. How can I set the `parallel` [parameter](https://ergo-code.github.io/HiGHS/dev/options/definitions/#parallel) in the above Python script? For ipm, is it off by default?

Call

`setOptionValue("parallel", "on")`

However the IPM solver is serial

>     3. How can I set `run_crossover` [parameter](https://ergo-code.github.io/HiGHS/dev/options/definitions/#run_crossover) to off in the above Python script? I see the crossover is set to on as default.

Call

`setOptionValue("run_crossover", "off")`

>     4. How can I set `infinite_cost` [parameter](https://ergo-code.github.io/HiGHS/dev/options/definitions/#infinite_cost) and `infinite_bound` [parameter](https://ergo-code.github.io/HiGHS/dev/options/definitions/#infinite_bound) using Python?

Call

- `setOptionValue("infinite_cost", value)`
- `setOptionValue("infinite_bound", value)`

However, if you're wanting to do  this so that you can use costs or bounds greater than 1e15, I strongly advise you not to, as you increase the danger of the solver failing. You should rescale your variables and constraints so that the costs and bounds are rather less than 1e15. Remember that HiGHS will be solving your model using primal/dual tolerances of 1e-7 using double precision (16-digit arithmetic) arithmetic, so costs/bounds greater than 1e16*1e-7=1e9 are asking for trouble

**Reply by hbshrestha** [reply]
Thank you very much! The `setOptionValue` [method](https://ergo-code.github.io/HiGHS/dev/interfaces/python/example-py/#example-py-option-values) worked for me to set the different solver options/settings.

Regarding scaling the variables and constraints, I understand it is the job of the user. Please flag if there are any parameters in HIGHS that can do such scaling.

**Reply by jajhall** [reply]
If you try to solve a badly-scaled problem, you'll see output like this

> Coefficient ranges:
  Matrix [1e+00, 3e+00]
  Cost   [1e+00, 1e+15]
  Bound  [0e+00, 0e+00]
  RHS    [8e+00, 1e+16]
WARNING: Problem has excessively large costs: consider scaling the costs by 1e+5 or less, or setting option user_cost_scale to -17 or less
WARNING: Problem has excessively large bounds: consider scaling the bounds by 1e+7 or less, or setting option user_bound_scale to -21 or less

If you don't want to do your own scaling, you can call

- setOptionValue("user_cost_scale", -17)
- setOptionValue("user_bound_scale", -21)

This will scale the _all_ the costs down by `2^{17}`, and _all_ the variable and constraint bounds down by `2^{21}`, possibly over-scaling some well-scaled variables and constraints, so that model precision is lost. [Some variables may be constrained to lie between bounds that are comparable with the primal feasibility tolerance, or some costs may be small comparable with the dual feasibility tolerance.

**Reply by hbshrestha** [reply]
Thank you for your response!

**Reply by hbshrestha** [reply]
I have started a new related discussion in a separate [thread](https://github.com/ERGO-Code/HiGHS/discussions/2052). I will close this one.

**Comment by huanxi0319** [comment]
Sure, I can help you with your questions regarding the HIGHS solver and its usage in Python. Let's address each of your questions one by one.

 1. Using the Interior Point Method (IPM)

The default method for solving LP problems in HIGHS is the simplex method. If you want to use the interior point method (IPM), you need to set the appropriate parameter. The `h.run(solver = "ipm")` syntax is not correct. Instead, you should use the `setOption` method to specify the solver. Here's how you can do it:

```python
h.setOption('solver', 'ipm')
```

After setting this option, you can call `h.run()` as usual. 
 2. Setting Parameters

To set parameters in HIGHS, you can use the `setOption**Value**` method. For example, if you want to enable parallelism, you would do something like this:

```python
h.setOptionValue('parallel', 'on')
```

If you want to check if a particular option is on or off, you can use the `getOptionValue` method. For instance, to check if the parallel option is enabled, you could do:

```python
print(h.getOptionValue('parallel'))
```

3. Disabling the Crossover

If you want to disable the crossover, which is used after the IPM to obtain a basic feasible solution, you can set the `run_crossover` option to `off`:

```python
h.setOptionValue('run_crossover', 'off')
```

 4. Setting Infinite Cost and Bound Parameters

To set the infinite cost and bound parameters, you can use the `setOptionValue` method as well. These parameters are typically used to define what HIGHS considers as an infinite value for costs and bounds. Here's how you can set them:

```python
# Set the infinite cost
h.setOptionValue('infinite_cost', 1e+20)  # Example value, adjust as needed

# Set the infinite bound
h.setOptionValue('infinite_bound', 1e+20)  # Example value, adjust as needed
```

### Putting It All Together

Here's how your script might look with the above changes:

```python
import highspy
import time

h = highspy.Highs()

# Read model
filename = 'abc.mps'
h.readModel(filename)

# Set options
h.setOptionValue('solver', 'ipm')  # Use the interior point method
h.setOptionValue('parallel', 'on')  # Enable parallel processing
h.setOptionValue('run_crossover', 'off')  # Disable crossover
h.setOptionValue('infinite_cost', 1e+20)  # Set infinite cost
h.setOptionValue('infinite_bound', 1e+20)  # Set infinite bound

# Run the model
h.run()
print('Model ', filename, ' has status ', h.getModelStatus())

solution = h.getSolution()
basis = h.getBasis()
info = h.getInfo()
model_status = h.getModelStatus()
print('Model status = ', h.modelStatusToString(model_status))
print()
print('Optimal objective = ', info.objective_function_value)
print('Iteration count = ', info.simplex_iteration_count)
print('Primal solution status = ', h.solutionStatusToString(info.primal_solution_status))
print('Dual solution status = ', h.solutionStatusToString(info.dual_solution_status))
print('Basis validity = ', h.basisValidityToString(info.basis_validity))
print(h.getOptionValue('parallel'))

# Write solution
h.writeSolution("abc_solution.sol", 1)
```

Make sure to replace the example values for `infinite_cost` and `infinite_bound` with the actual values that make sense for your problem, although setting these to large values is unwise.

**Reply by jajhall** [reply]
Sorry @huanxi0319 , but there are errors in your reply. In particular

* The IPM solver is available in all builds of HiGHS. Indeed, in terms of what's available, there is no difference between builds 

* Option values are set by calling `setOptionValue`, not `setOption`. Ditto, it's  `getOptionValue`, not `getOption` to get an option value

It's unfortunate that you didn't check your Python code before posting it here 

I've correct the errors in your reply
